internal/controller/imagerepository_controller.go

```diff
@@ -877,3 +879,222 @@ func (r *ImageRepositoryReconciler) ImageRepositoryForSameUrlExists(ctx context.
+		log.Error(err, "failed to update application pull secret", "secretName", applicationPullSecretName)
+		return err
+	}
+
+	log.Info("Application pull secret updated with new registry auth", "application", applicationName)
+	return nil
+
+}
+
+func (r *ImageRepositoryReconciler) removePullSecretFromApplicationPullSecret(ctx context.Context, applicationName string, namespace string, pullSecretName string, registryUrl string) error {
+	log := ctrllog.FromContext(ctx)
+
+	applicationPullSecretName := getApplicationPullSecretName(applicationName)
+	applicationPullSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: applicationPullSecretName, Namespace: namespace}, applicationPullSecret); err != nil {
+		log.Error(err, "failed to get application pull secret", "secretName", applicationPullSecretName)
+		return err
+	}
+
+	// Get the pull secret that is being removed
+	pullSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: pullSecretName, Namespace: namespace}, pullSecret); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("Pull secret already deleted, cannot identify which registry auths to remove", "secretName", pullSecretName)
+			return nil
+		}
+		log.Error(err, "failed to get pull secret", "secretName", pullSecretName)
+		return err
+	}
+	if pullSecret.Type != corev1.SecretTypeDockerConfigJson {
+		log.Info("Skipping secret due to unexpected type", "secretName", pullSecretName, "type", pullSecret.Type)
+		return nil
+	}
+
+	dockerConfigDataBytes, ok := pullSecret.Data[corev1.DockerConfigJsonKey]
+	if !ok {
+		log.Info("Missing .dockerconfigjson key", "secretName", pullSecretName)
+		return nil
+	}
+
+	var toRemove dockerConfigJson
+	if err := json.Unmarshal(dockerConfigDataBytes, &toRemove); err != nil {
+		log.Error(err, "failed to unmarshal .dockerconfigjson", "secretName", pullSecretName)
+		return err
+	}
+
+	var existing dockerConfigJson
+	if err := json.Unmarshal(applicationPullSecret.Data[corev1.DockerConfigJsonKey], &existing); err != nil {
+		log.Error(err, "failed to unmarshal application .dockerconfigjson", "secretName", applicationPullSecretName)
+		return err
+	}
+
+	if existing.Auths == nil {
+		return nil
+	}
+
+	changed := false
+	for reg := range toRemove.Auths {
+		// Check if there’s another IR with the same repo URL. In that case
+		// we don't remove the record for the registry pullsecret completely, but replace it
+		// with pullsecret from this other IR.
+		foundImageRepositoryWithSameUrl := false
+
+		imageRepositoriesList := &imagerepositoryv1alpha1.ImageRepositoryList{}
+		if err := r.Client.List(ctx, imageRepositoriesList, &client.ListOptions{Namespace: namespace}); err != nil {
+			log.Error(err, "failed to list image repositories")
+			return err
+		}
+
+		for _, otherIR := range imageRepositoriesList.Items {
+			// Skip the IR with the pullsecret we're deleting
+			if otherIR.Status.Credentials.PullSecretName == pullSecretName {
```

Comment:
I'd rather pass to method CURRENT imageRepository name and check that

Reply:
Don't understand this comment.

Reply:
ah, I understand now.